src/zkp/piaffg.rs

```diff
@@ -204,143 +361,173 @@ impl Proof for PiAffgProof {
         context: &impl ProofContext,
         transcript: &mut Transcript,
     ) -> Result<()> {
-        Self::fill_transcript(
-            transcript, context, input, &self.S, &self.T, &self.A, &self.B_x, &self.B_y, &self.E,
-            &self.F,
+        // Generate verifier's challenge via Fiat-Shamir...
+        let challenge = Self::generate_challenge(
+            transcript,
+            context,
+            input,
+            &self.mult_coeff_commit,
+            &self.add_coeff_commit,
+            &self.random_affine_ciphertext,
+            &self.random_mult_coeff_exp,
+            &self.random_add_coeff_ciphertext_1,
+            &self.random_mult_coeff_commit,
+            &self.random_add_coeff_commit,
         )?;
-        // Verifier samples e in +- q (where q is the group order)
-        let e = plusminus_bn_random_from_transcript(transcript, &k256_order());
-
-        if e != self.e {
+        // ... and check that it's the correct challenge.
+        if challenge != self.challenge {
             warn!("Fiat-Shamir consistency check failed");
             return Err(InternalError::ProtocolError);
         }
-
-        // Do equality checks
-
-        let eq_check_1 = {
-            let a = input
-                .pk0
-                .encrypt_with_nonce(&self.z2, &self.w)
-                .map_err(|_| InternalError::ProtocolError)?;
+        // Check that the affine-like transformation holds over the masked
+        // coefficients using the 0th encryption key.
+        let masked_affine_operation_is_valid = {
+            let tmp = input
+                .encryption_key_0
+                .encrypt_with_nonce(&self.masked_add_coeff, &self.masked_add_coeff_nonce_0)
+                .map_err(|_| InternalError::InternalInvariantFailed)?;
             let lhs = input
-                .pk0
-                .multiply_and_add(&self.z1, &input.C, &a)
-                .map_err(|_| InternalError::ProtocolError)?;
+                .encryption_key_0
+                .multiply_and_add(&self.masked_mult_coeff, &input.input_ciphertext, &tmp)
+                .map_err(|_| InternalError::InternalInvariantFailed)?;
             let rhs = input
-                .pk0
-                .multiply_and_add(&self.e, &input.D, &self.A)
-                .map_err(|_| InternalError::ProtocolError)?;
+                .encryption_key_0
+                .multiply_and_add(
+                    &self.challenge,
+                    &input.output_ciphertext,
+                    &self.random_affine_ciphertext,
+                )
+                .map_err(|_| InternalError::InternalInvariantFailed)?;
             lhs == rhs
         };
```

Comment:
suggestion: I was wondering if this would be more readable if we collect the errors via a closure and do one `map_err` at the end, but the more complicated machinery might not be worth it. E.g.
```rust 
let masked_affine_operation_is_valid = || -> std::result::Result<bool, PaillierError> {
    let tmp = input
        .encryption_key_0
        .encrypt_with_nonce(&self.masked_add_coeff, &self.masked_add_coeff_nonce_0)?;
    let lhs = input
        .encryption_key_0
        .multiply_and_add(&self.masked_mult_coeff, &input.input_ciphertext, &tmp)?;
    let rhs = input
        .encryption_key_0
        .multiply_and_add(
             &self.challenge,
             &input.output_ciphertext,
             &self.random_affine_ciphertext,
        )?;
    Ok(lhs == rhs)
}()
.map_err(|_| InternalError::InternalInvariantFailed)?;
```